Added decode option to config set command

Catalin-Ionut opened the pull request:
### 1. Why is this change necessary?
#2207 
### 2. What does this change do, exactly?
Added `decode` input option to the `system:config:set` command.
### 3. Describe each step to reproduce the issue or behaviour.
#2207 
### 4. Please link to the relevant issues (if any).
#2207 
### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

Catalin-Ionut committed:
Added decode option to config set command

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/platform?pullRequest=2209) <br/>All committers have signed the CLA.

shyim commented on src/Core/System/SystemConfig/Command/ConfigSet.php:
```diff
@@ -29,6 +29,8 @@ protected function configure(): void
             ->addArgument('key', InputArgument::REQUIRED)
             ->addArgument('value', InputArgument::REQUIRED)
+            /* @deprecated tag:v6.5.0 - Will be made REQUIRED in v6.5.0 */
+            ->addArgument('type', InputArgument::OPTIONAL, 'string|bool|boolean', 'string')
```
Maybe we could allow also json. It could theoretically also solve the bool conversation problem. What do you think

keulinho replied:
we also should add `int` and `float`

mitelg replied:
will post again, how it was solved in SW5 :grin: https://github.com/shopware/shopware/blob/5.7/engine/Shopware/Commands/ConfigSetCommand.php#L61-L63

if decode is used, it will automatically be the right type

keulinho commented:
Please also add testcases for the different type encodings.

Catalin-Ionut commented:
> Please also add testcases for the different type encodings.

I added the tests and changed the implementation to json_decode that handles multiple cases.

keulinho commented on src/Core/System/Test/SystemConfig/Command/ConfigSetCommandTest.php:
```diff
@@ -0,0 +1,70 @@
+        ];
+
+        if ($decode) {
+            $command += ['--decode' => true];
```
```suggestion
            $command['--decode'] = true;
```
you don't need the `+=` operator, we try to avoid it when using arrays as it can be misleading

keulinho commented on src/Core/System/Test/SystemConfig/Command/ConfigSetCommandTest.php:
```diff
@@ -0,0 +1,70 @@
+    public function configValueProvider(): iterable
+    {
+        /* config_key, value, expected_value, decode */
+        yield 'String false' => ['fake_config_key', 'false', 'false', false];
```
the `key` value is always the same, you can hard code it in the testCase itself, that way the test data provider get's denser and is easier to understand

keulinho commented on src/Core/System/Test/SystemConfig/Command/ConfigSetCommandTest.php:
```diff
@@ -0,0 +1,70 @@
+        yield 'String int' => ['fake_config_key', '4', '4', false];
+        yield 'Decode String int' => ['fake_config_key', '5', 5, false];
+        yield 'String float' => ['fake_config_key', '2.2', '2.2', false];
+        yield 'Decode String float' => ['fake_config_key', '3.3', 3.3, false];
```
```suggestion
        yield 'Decode String float' => ['fake_config_key', '3.3', 3.3, true];
```
i didn't execute the tests, but this one should have failed right?

Catalin-Ionut replied:
As i have found out the ->with(...) uses IsEqual so it does not check type, i changed the test to check type manually with a callback.

keulinho replied:
as described further down you should be able to use `isIdenticalTo(..)`

keulinho commented:
I have just some minor suggestions, besides that the PR looks good to me :+1:

keulinho commented on src/Core/System/Test/SystemConfig/Command/ConfigSetCommandTest.php:
```diff
@@ -0,0 +1,70 @@
+            "{\"name\":\"abc\",\"place\":\"xyz\"}",
+            false,
+        ];
+        yield 'Decode String json' => [
```
i also would add a test case that some random string will be passed as is even if the decode option is enabled
```php
yield 'Decode string remains string' => [
    'random string',
    'random string',
     true,
];
```

Catalin-Ionut replied:
Done.

keulinho commented on src/Core/System/SystemConfig/Command/ConfigSet.php:
```diff
@@ -30,17 +30,31 @@ protected function configure(): void
+    {
+        $value = $input->getArgument('value');
+        if ($input->getOption('decode')) {
+            $decodedValue = json_decode($value, true);
```
```suggestion
            return json_decode($value, true, 512, \JSON_THROW_ON_ERROR);
```
by using the Throw on error flag you can get rid of the manual check to the json erros

and besides that i would prefer throwing an exception in the case that the decode option is used, but the provided value can't be decoded instead of silencing the error and continuing with the raw value

Can you also add a test case for invalid json?

Catalin-Ionut replied:
if i use the `JSON_THROW_ON_ERROR` decoding a simple string will throw error so the test case
`yield 'Decode string remains string' => ['random string', 'random string', true];` will not work anymore since it will throw error.

keulinho commented on src/Core/System/Test/SystemConfig/Command/ConfigSetCommandTest.php:
```diff
@@ -0,0 +1,77 @@
+
+        $this->systemConfigService->expects($this->once())
+            ->method('set')
+            ->with(
```
Does
```suggestion
            ->with($key, static::identicalTo($expectedValue), TestDefaults::SALES_CHANNEL);
```
work? with the `identicalTo()` constraint phpunit should perform a type safe comparisons

if that does not work as expected i'm fine with your solution of using an callback ;)

keulinho commented:
LGTM :+1:

marcelbrode commented:
Thanks for your contribution! :blue_heart: :tada:
